### PR TITLE
Draft of generic impl

### PR DESCRIPTION
**Pull request by tcoratger**

### CODE REVIEW

File: crates/evm/src/lib.rs [hunk 1]
```diff
@@ -23,6 +24,18 @@ pub mod provider;
...
 pub mod test_utils;
 
+/// Trait for configuring the EVM in a generic way
+pub trait ConfigureEvmGeneric: ConfigureEvm + ConfigureEvmEnv {
+    /// Returns a new EVM with the given database configured with the given environment settings.
+    fn evm_with_env_ext<'a, DB>(
+        &'a self,
+        db: DB,
+        env: EnvWithHandlerCfg,
+    ) -> Box<dyn EvmTransact<DB = DB> + 'a>
```
Comment (greged93) [inline crates/evm/src/lib.rs:34]: also do we have to return a trait object? Can't you add a generic `EVM: EvmTransact<DB = DB>`?

File: crates/evm/src/execute.rs [hunk 2]
```diff
@@ -14,33 +19,123 @@ pub use reth_storage_errors::provider::ProviderError;
...
-    type Output;
-    /// The error produced by the evm.
-    type Error;
+    /// The database used by the evm.
+    type DB: Database;
 
     /// Transacts with the current env to produce an output without
     /// committing to the underlying database.
-    fn transact(&mut self) -> Result<Self::Output, Self::Error>;
+    fn transact(&mut self) -> EVMResult<<<Self as EvmTransact>::DB as Database>::Error>;
```
Comment (greged93) [inline crates/evm/src/execute.rs:23]: I'm not sure we want to tie the error here to always be DB::Error

Comment (greged93) [reply crates/evm/src/execute.rs:23]: Is it an issue to have:
```
    /// The output produced by the evm.
    type Output;
    /// The error produced by the evm.
    type Error;
```
?

File: crates/evm/src/execute.rs [hunk 3]
```diff
@@ -14,33 +19,123 @@ pub use reth_storage_errors::provider::ProviderError;
...
+    fn db_mut(&mut self) -> &mut Self::DB;
+
+    /// Returns the reference of transaction.
+    fn tx(&self) -> &TxEnv;
+
+    /// Returns the mutable reference of transaction.
+    fn tx_mut(&mut self) -> &mut TxEnv;
+
+    /// Returns the reference of block.
+    fn block(&self) -> &BlockEnv;
```
Comment (greged93) [inline crates/evm/src/execute.rs]: I don't know if having those getters in the `EvmTransact` trait makes sense

File: crates/evm/src/execute.rs [hunk 4]
```diff
@@ -14,33 +19,123 @@ pub use reth_storage_errors::provider::ProviderError;
...
+}
+
+/// Trait that represents the output of a transaction and provides access to the state.
+pub trait EvmTransactOutput {
+    /// Returns the state produced by the transaction.
+    fn state(&self) -> EvmState;
 }
```
Comment (greged93) [inline crates/evm/src/execute.rs]: Why is this needed?

File: crates/evm/src/execute.rs [hunk 5]
```diff
@@ -14,33 +19,123 @@ pub use reth_storage_errors::provider::ProviderError;
...
 /// commit state changes to the database.
 pub trait EvmCommit: EvmTransact {
+    /// The output produced by the evm.
+    type CommitOutput;
+    /// The error produced by the evm.
+    type CommitError;
+
     /// Commit state changes to the database.
-    fn commit(&mut self, output: Self::Output);
+    fn commit(&mut self, output: HashMap<Address, Account>);
```
Comment (greged93) [inline crates/evm/src/execute.rs]: this should be `Self::CommitOutput` no?

File: crates/evm/src/execute.rs [hunk 6]
```diff
@@ -14,33 +19,123 @@ pub use reth_storage_errors::provider::ProviderError;
...
+pub trait EvmTransactOutput {
+    /// Returns the state produced by the transaction.
+    fn state(&self) -> EvmState;
 }
 
 /// Trait that can transact an [`EvmTransact::Env`] and
 /// commit state changes to the database.
 pub trait EvmCommit: EvmTransact {
+    /// The output produced by the evm.
+    type CommitOutput;
```
Comment (greged93) [inline crates/evm/src/execute.rs]: maybe we should rename this `EvmOutput`?

File: crates/evm/src/execute.rs (hunk 2 above)
Comment (tcoratger) [reply crates/evm/src/execute.rs:23]: I think that for a first iteration it could be a problem yes because all the returns in reth are configured like that and therefore it would change a lot of things. I figured this was something we could make more generic in the future on other iterations.

File: crates/evm/src/execute.rs (hunk 3 above)
Comment (tcoratger) [reply crates/evm/src/execute.rs]: I've removed the block and tx related ones, but in my opinion `env_with_handler_cfg` is needed for this iteration since we have

https://github.com/paradigmxyz/reth/blob/9210269d3aea19a25ea9530399f98d7dc831eef8/crates/rpc/rpc-eth-api/src/helpers/call.rs#L303

File: crates/evm/src/execute.rs (hunk 4 above)
Comment (tcoratger) [reply crates/evm/src/execute.rs]: not needed for this iteration, just removed

File: crates/evm/src/execute.rs (hunk 5 above)
Comment (tcoratger) [reply crates/evm/src/execute.rs]: I do not think so. I just changed the naming because it was indeed a bit confusing but the output of commit is not the same as the output of transact and commit.

Indeed in commit, I think we can say that in a fairly generic way (true for any implementation of evm), the output must be a hashmap of address and account after the commit.

On the other hand, the output of transact and commit is an `ExecutionResult` (success, revert, halt).

File: crates/evm/src/execute.rs (hunk 6 above)
Comment (tcoratger) [reply crates/evm/src/execute.rs]: agreed, just changed

File: crates/evm/src/lib.rs (hunk 1 above)
Comment (tcoratger) [reply crates/evm/src/lib.rs:34]: In this situation there are inference problems if we choose the approach you mention and all the compiler's suggestions converge towards the use of a trait object I have the impression...

File: crates/evm/src/execute.rs (hunk 5 above)
Comment (greged93) [reply crates/evm/src/execute.rs]: but you could argue that the commit can take the evm transact output `ResultAndState` and ingest this in order to commit the state changes to the db. This makes the function more generic and we already have the `ResultAndState` struct.

Comment (tcoratger) [reply crates/evm/src/execute.rs]: You are right on this. I just used the `ResultAndState` as the commit argument. But I didn't add a generic type for TransactOutput. It added one more generic and a little complexity. Let's keep this for a follow up in order to restrict the scope as much as possible here knowing that it is not hyper blocking.
